Make DoFRenumbering::component_wise() work with FEValuesExtractors

gsegon opened the pull request:
Implements #17719.  

Implementation using std::variant to group supported FEValuesExtractors structs into a convenience alias DoFRenumbering::ExtractorVariant. The implementation is sanity-checked with step-22, no unit-test are introduced at the moment. Alternative approaches aimed at generalization and getting rid of std::holds_alternative checks proved to be not worth the complexity introduced, at least without sub-classing extractors from a common base which I assume is not wanted. If there's a better approach please do let me know. 

Kindly requesting general feedback and in case you'd proceed with this PR a suggestion on how to go about unit-testing this.

~G

gsegon committed:
Implement `component_wise` renumbering supporting FEValuesExtractors for component grouping.
Applied review suggestions.

bangerth commented on include/deal.II/dofs/dof_renumbering.h:
```diff
@@ -689,6 +690,76 @@ namespace DoFRenumbering
 
+
+
+  /**
+   * Convenience alias for variant type of supported FEValuesExtractors structs
+   */
+  using ExtractorVariant = std::variant<FEValuesExtractors::Scalar,
+                                        FEValuesExtractors::Vector,
+                                        FEValuesExtractors::Tensor<2>,
+                                        FEValuesExtractors::SymmetricTensor<2>>;
```
The use of a `std::variant` here is clever -- I did not think of this in other places. We may want to encourage this in other places. I think this could perhaps move into the header files where we declare the extractors.

bangerth commented on include/deal.II/dofs/dof_renumbering.h:
```diff
@@ -689,6 +690,76 @@ namespace DoFRenumbering
+                                        FEValuesExtractors::SymmetricTensor<2>>;
+
+
+  /**
+   * Utility function used to fill a given `component_order` vector by
+   * processing an `order` of FEValuesExtractors. Supported extractors are
+   * listed in DoFRenumbering::ExtractorVariant. The function is used in the
+   * implementation of component_wise(DoFHandler<dim, spacedim>&, const
+   * std::vector<ExtractorVariant> &).
+   */
+  template <int dim, int spacedim>
+  void
+  populate_component_order(
+    const std::vector<ExtractorVariant> &order,
+    std::vector<unsigned int>           &component_order,
+    unsigned int unassigned_value = numbers::invalid_unsigned_int);
```
If it's a utility function, just move it into the `.cc` file as a function in an anonymous namespace without declaring it here. It should only be declared in the `.h` file if you think it's useful to users.

bangerth reviewed (commented):
I'll take a closer look at this after I'm back from vacation late next week, unless someone else beats me to it.

bangerth commented:
Also, nice job tackling this issue! We love it when people take a look at open issues and help out!

bangerth reviewed (changes requested):
I like this. I have a few comments, but none of them is important. My apologies for being slow to get back to this!

Let me know once you've got these comments addressed!

gsegon commented on examples/step-22/step-22.cc:
```diff
@@ -461,7 +461,21 @@ namespace Step22
+    // call DoFRenumbering::component_wise with a braced initializer list of
+    // extractors. For details, see DoFRenumbering::component_wise<dim,
+    // spacedim>(DoFHandler<dim, spacedim> &, const
+    // std::vector<ExtractorVariant> &).
```
Done. reverted to version from master. No step-22.cc changes are in contained PR.

gsegon commented on include/deal.II/dofs/dof_renumbering.h:
```diff
@@ -689,6 +689,53 @@ namespace DoFRenumbering
+   * Calling the function with {velocities, pressure} is equivalent to calling
+   * the above function with std::vector<unsigned int>{0, 0, 1} for dim=2, while
+   * calling with {pressure, velocities} would result in a call to above
+   * function with std::vector<unsigned int>{1, 1, 0} as the target_component
```
Done.

gsegon commented on include/deal.II/dofs/dof_renumbering.h:
```diff
@@ -689,6 +689,53 @@ namespace DoFRenumbering
+  /**
+   * Sort the degrees of freedom by vector component. It does the same thing as
+   * above function except the ordering argument is a braced initializer list
+   * containing FEValuesExtractors.
```
Done.

gsegon commented on include/deal.II/dofs/dof_renumbering.h:
```diff
@@ -689,6 +689,53 @@ namespace DoFRenumbering
+   * @endcode
+   *
+   * Calling the function with {velocities, pressure} is equivalent to calling
+   * the above function with std::vector<unsigned int>{0, 0, 1} for dim=2, while
```
Done.

gsegon commented on include/deal.II/fe/fe_values_extractors.h:
```diff
@@ -309,6 +310,14 @@ namespace FEValuesExtractors
+  using ExtractorVariant = std::variant<FEValuesExtractors::Scalar,
+                                        FEValuesExtractors::Vector,
+                                        FEValuesExtractors::Tensor<2>,
+                                        FEValuesExtractors::SymmetricTensor<2>>;
```
Done.

gsegon commented on source/dofs/dof_renumbering.cc:
```diff
@@ -746,6 +746,174 @@ namespace DoFRenumbering
+    void
+    populate_component_order(
+      const std::vector<FEValuesExtractors::ExtractorVariant> &order,
+      std::vector<unsigned int>                               &component_order,
```
Rewritten such that `component_order` is returned by the function and not passed in as a reference. The function is also renamed to `generate_component_order`. The function arguments are now a const ref vector of FEValueExtractors and  `fe_n_components`. 

Internally, the `fe_n_components` is used to pre-allocate the `component_order` (std::vector<unsigned int>) to correct size which is later returned. On the caller site of `generate_component_order`, `dof_handler.get_fe().n_components()` is passed as the `fe_n_components`.

Perhaps `fe_system_components` for  `fe_n_components` could be a better name.

gsegon commented on source/dofs/dof_renumbering.cc:
```diff
@@ -746,6 +746,174 @@ namespace DoFRenumbering
+    populate_component_order(
+      const std::vector<FEValuesExtractors::ExtractorVariant> &order,
+      std::vector<unsigned int>                               &component_order,
+      unsigned int unassigned_value = numbers::invalid_unsigned_int)
```
Yes, the initial intention was to prepare the `component_order` with correct size and initialize with `unassigned_value` to guard against overlapping assignments while populating (i.e. only assign if not assigned before). 

Now, the function is rewritten as discussed in the previous comment so now the size is passed in to construct the `component_order` to correct size and values are initialized to `numbers::invalid_unsigned_int`. `unassigned_value` was hardcoded to `numbers::invalid_unsigned_int`, and the argument removed.

gsegon commented on source/dofs/dof_renumbering.cc:
```diff
@@ -746,6 +746,174 @@ namespace DoFRenumbering
+              num_components = 1;
+            }
+
+          else if (std::holds_alternative<FEValuesExtractors::Vector>(
```
Done.

gsegon commented on source/dofs/dof_renumbering.cc:
```diff
@@ -746,6 +746,174 @@ namespace DoFRenumbering
+        "numbers::invalid_unsigned_int must be present after processing all given "
+        "extractors. This error typically occurs when the component_wise function "
+        "is passed fewer extractors than needed or when the set of extractors "
+        "provided in the extractor_order argument doesn't cover all required components. "));
```
Added 
"   ...
   * The initializer list of extractors must cover all finite-element
   * components, and each component must be covered by exactly one extractor.
   * Any missing or duplicate assignment to finite-element components will
   * produce an error.
 "
to function documentation in `dof_renumbering.h`.

luca-heltai reviewed (approved):
I like this! Thanks!

bangerth commented on include/deal.II/dofs/dof_renumbering.h:
```diff
@@ -689,6 +689,58 @@ namespace DoFRenumbering
+   * Calling the function with `{velocities, pressure}` is equivalent to calling
+   * the above function with `std::vector<unsigned int>{0, 0, 1}` for dim=2,
+   * while calling with `{pressure, velocities}` would result in a call to above
+   * function with `std::vector<unsigned int>{1, 1, 0}` as the
```
Actually, I think it is `{1,0,0}` in the second case -- `pressure` corresponds to a scalar.

luca-heltai replied:
@bangerth, I think it is correct as it is. The internal numbering depends on the arguments passed to the extractors. So the call would result in two blocks, where the first correspond to the scalar extractor (with index 0), and the second to the vector extractor (with index 1). In the FESystem, however, they are still ordered as in the constructor of FESystem, i.e., vector first, scalar later. Unless you used a `Scalar(0)` and a `Velocity(1)` constructors, the result is what is indicated in the documentation...

gsegon replied:
@bangerth, sure will squash before merging, was expecting the comment on this anyway :). The code as implemented will provide {1, 1, 0} and the documentation doesn't have a typo on this part. Although I'm double checking the logic and perhaps add further notes. The documentation as is might be misleading without mentioning what @luca-heltai comments, imo.

@luca-heltai That is my understanding as well. Although, I'm doing some sanity checks now. 

Basically, in step-22

If `fe(FE_Q<dim>(degree + 1) ^ dim, FE_Q<dim>(degree))` then 
- `{0, 0, 1}` produced by `{velocities, pressure}` will work as expected.
- `{1, 1, 0}` produced by `{pressure, velocities}` produces an error from UMFPACK reporting that matrix is singular but factorization successful anyway.

If `fe(FE_Q<dim>(degree), FE_Q<dim>(degree + 1) ^ dim)`, with accompanying changes in the code, then:
- `{0, 1, 1}` produced by `{pressure, velocities}` works as expected.
- `{1, 0, 0}` produced by `{velocities, pressure}`  produces an error from UMFPACK reporting that matrix is singular but factorization successful anyway.

What I'm unsure of though is if the cases where UMFPACK produces the error make sense in the first place for a user to do such reordering or should this be also guarded against? I'd like to take a bit more time to clear this for myself...

bangerth replied:
@luca-heltai Yes, you are right, thanks for pointing it out!

The issue with UMFPACK has to do what block(s) of the matrix UMFPACK tries to compute an inverse of. If that involves the bottom right block, we shouldn't be surprised about the result. In any case, I think it's all fine -- just squash the commits and let me know when you're done!

gsegon replied:
@bangerth  Squashed and pushed, done.

@luca-heltai Yes, it is not obvious to me where the issue is but I do think it's something trivial arising from problem setup in those case and not from a bug from the library. When I find time I'll have another look to understand it.

Side note: I've done some refactoring in a [separate branch ](https://github.com/dealii/dealii/compare/master...gsegon:dealii:dofrenumbering-component_wise-extractors-config-sanity-check#diff-fdb30f4a2a85662c2593ed3ce3c01bfdfca62c8e6dc4b9988110aca0a1d50d49) to understand how `compute_component_wise()` works as I expected that `{0, 0, 2}` would also be a valid input and had some confusion about how its values correspond to fe component id's. They don't - as expected, the values are only used for grouping but using `{0, 0, 2}` was anyway not valid. It could be valid, as implemented in the branch. This was an exercise to confirm that the values of `target_component` are not tied in anyway to fe component indices. The _indices_ of `target_component` correspond to component indices of the fe system. _Values_ of `target_component` are used to group dofs.   `compute_component_wise()` _can_ be made to work with input `{0, 0, 2}`, or `{7, 7, 15}`, producing the same result as `{0, 0, 1}`, although no functionality is lost without it.

bangerth reviewed (changes requested):
I found one more minor thing. Would you mind
* Addressing the one small issue
* Squashing or merging all of your commits into one?
With that, I think this is ready!

bangerth commented:
It looks like docker is down along with the rest of the internet today :-( 

@gsegon If you have further experiments that result in small test cases you are playing with, a good strategy is to turn these into additions to the test suite. In fact, that's how I usually operate: When I play with something, I almost always do this with programs I add to the test suite locally (and then often submit them as patches). There is no such thing as too many tests -- if you have 7 small programs that cover different ways of setting up arguments that all turns out to be valid, then these might as well become 7 more tests in the test suite!

bangerth commented:
As you mention, though, just because something *can* be made to work doesn't mean that it *should* be made to work. Oftentimes it is useful to only allow the simplest choice -- say arrays such as `{0,1,2}`, but not `{0,2,17}` if the only thing that matters is the relative order of array entries: It's true that the latter has the same semantic content, but my preference would be to standardize on the former anyway.

bangerth commented:
OK, that sounds good. It shouldn't actually be very difficult to adapt one of the existing tests. For example, you could clone the `tests/dofs/dof_renumbering.cc` test that already uses the function in question with an ordering argument. Just duplicate this code (and remove the stuff that happens in that test after the lines that do
```
  DoFRenumbering::component_wise(dof, order);
  print_dofs(dof);
```
with the exception of the fact that it also calls the function on the MG dofs which we want to continue testing. There are many other tests you can find via
```
grep component_wise tests/*/*cc
```
and looking specifically for the ones that use the second argument of the `component_wise()` function.

Let us know if you'd like help and/or are at a point where you want us to look over the patch again!

gsegon commented:
@bangerth thanks for the pointers - will do, probably this weekend, and let you guys know when done.

bangerth commented on include/deal.II/fe/fe_values_extractors.h:
```diff
@@ -309,6 +310,16 @@ namespace FEValuesExtractors
+  using ExtractorVariant = std::variant<FEValuesExtractors::Scalar,
+                                        FEValuesExtractors::Vector,
+                                        FEValuesExtractors::Tensor<2>,
+                                        FEValuesExtractors::SymmetricTensor<2>>;
```
Since you were going to make more changes to this PR, here's one other thing: I had second thoughts about the name of this declaration. I would name it `AnyExtractor`, I think that expresses the purpose better.

gsegon replied:
Done.

gsegon committed:
indent-all
Swapped test 12 and 13 names
Simplify output of DoF renumbering test by removing unnecessary details.
Add detailed DoF ordering checks to test and update expected output.
Updated test and expected output.
Rename `ExtractorVariant` to `AnyExtractor` and add test output for DoF renumbering verification.

gsegon commented on tests/dofs/dof_renumbering_12.output:
Purposely identical to `dof_renumbering.output`.

bangerth commented:
Let us know if you're at a point where you want someone to take another look!

gsegon committed:
Added a test case using all supported AnyExtractors.
Added a bound check for `component_order` element access.
Added invalid input checks.
Added MG checks.

gsegon commented:
Ok, please review :).

Also, just to confirm, we want the components of a given extractor to be block together, right? This results in global dof bounds of these components (per-extractor components) to overlap. 
Is there any case where a user might not want them blocked together but nevertheless ordered using extractors? In this case their global dof bounds would not overlap.

bangerth commented:
/rebuild

bangerth reviewed (approved):
My apologies for dropping the ball on this PR -- this looks good and I really like the new functionality!

bangerth commented:
I'll get this merged as soon as the testers are all green. If you wouldn't mind, as a follow-up pull request, could you add an entry to `doc/news/changes/minor` that describes the new function?

gsegon commented:
> I'll get this merged as soon as the testers are all green. If you wouldn't mind, as a follow-up pull request, could you add an entry to `doc/news/changes/minor` that describes the new function?

Sure, will add. (Fri or Sat)
